### DISCUSSION TITLE
Submit is always successful even when the request is rejected

### DISCUSSION THREAD

**Discussion by ergunpp** [react-hook-form/react-hook-form#4012] [Q&A] [answered]
Hi, 
I have an async submit handler, which posts data to a server on submit. My code in onsubmit is:

```
    const onSubmit = async (data) => {
        const prodOrderData = {
            itemId: data.product[0].id,
            qty: data.qty,
            bomId: data.product[0].bom.id,
            revId: data.product[0].bom.revId,
            varId: data.variant && data.variant.length > 0 ? data.variant[0].id : undefined,
            planDate: data.beginDate.toJSON(),
            uomRef: data.product[0].units.find(item => item.isMainUnit === 1).untLnRef,
            createDate: data.date.toJSON(),
            hasVariant: data.product[0].hasVrnts

        }
        const request = prepareRequest(prodOrderData)
        const token = await getToken()
        let postData = new Promise((resolve,reject) => {
            
            if (!request.valid) {
                reject('Invalid data')
            }
            try {
                console.log(request)
            const response = axios.post(request.path, request.body, {
                headers: {
                    'Authorization': token,
                    
                }
            })
            console.log(response)
            resolve(response)
        } catch (error) {
            reject (error)
        }
        })
       

       const result = await postData
        
        return result


    }
```

What data do I have to return to catch when the submit is not successful?

**Comment by bluebill1049** [comment]
you can up vote this feature request: https://github.com/react-hook-form/react-hook-form/projects/1 at the moment we don't catch any error throw at handlSubmit callback.

**Comment by ergunpp** [comment]
So it means I have to handle it with general state myself?

**Reply by bluebill1049** [reply]
yes.

**Comment by santino** [comment, marked as answer]
You can still pass `setError` to your `onSubmit` function.
In this case rather than rejecting you can do something like:
```js
return setError('submitError', {
  type: 'server',
  message: 'Invalid data'
})
```

Notice that I'm setting a 'submitError', and not an error on a specific field in this case; which seems to be what you want.
When you do this, `isSubmitSuccessful` becomes false, because there is an error on the form; which hopefully is what you need.

In your component, you can have a `<FormSubmitError>` that displays the messages exposed by `formState.errors.submitError`.

With this setup, you don't have to maintain internal state and can still leverage React Hooks Form functionality.
